pkg/util/http.go

```diff
@@ -195,60 +198,73 @@ func (e MsgSizeTooLargeErr) Is(err error) bool {
 	return ok1 || ok2
 }
 
-func decompressRequest(dst []byte, reader io.Reader, expectedSize, maxSize int, compression CompressionType, sp opentracing.Span) (body []byte, err error) {
-	defer func() {
-		if err != nil && len(body) > maxSize {
-			err = MsgSizeTooLargeErr{Actual: len(body), Limit: maxSize}
-		}
-	}()
+func decompressRequest(dst []byte, reader io.Reader, expectedSize, maxSize int, compression CompressionType, sp opentracing.Span) ([]byte, error) {
```

Comment:
Thanks, will have a look tomorrow!

Reply:
Thanks for the revision @pstibrany - I incorporated it with some minor tweaks I thought made sense. PTAL :)

Reply:
Looks like the revision introduced a Snappy bug, let me try and solve it.

Reply:
I realize that `decompressSnappyFromBuffer` in the revision is problematic, since it's used also if the incoming reader can be converted to a buffer, _and is not compressed_. In that case, an attempt is made to decompress uncompressed input as snappy, which is why the test fails.

Reply:
`decompressFromBuffer` on its own is quite confusing, since it doesn't always do decompression. It really only works together with `decompressRequest`.

That's why I would suggest to only handle Snappy case in `decompress(Snappy)FromBuffer` function, and move handling of `NoCompression` and `Gzip` into `decompressRequest`.

We can easily handle `NoCompression` case in the same branch as conversion to buffer.

Reply:
OK thanks for your input @pstibrany, I'll handle `NoCompression`  in the conversion to buffer branch.

Reply:
I've implemented your suggestion to handle `NoCompression` outside of `decompressSnappyFromBuffer` @pstibrany. WDYT?